report/3.Theory/LanguageSpecification.tex

```diff
@@ -0,0 +1,192 @@
+
+\textbf{Operators}\\
+Most of the operators in the language are defined as functions.
+This means that all arithmetic operators are infix-functions that takes two numbers as parameters. 
+The only operators that remains operators are the assign-operator (\textbf{=}) and the encapsulation-operators, namely parenthesis, square-brackets and curly-brackets.
+
+Not having operators for arithmetic operations also means that there are no native operator precedence in the language.
+This can be amended by structuring calculations in a left-to-right manner, or by encapsulating operations with parenthesis. 
+
+\textbf{Overloading}\\
+
+The language will include function overloading. 
+This means that multiple functions can share the same name and have different implementations based on the different parameters that the function accepts. 
+An example could be the plus function. 
+The plus function could accept two numbers or two strings, but the implementation of the plus function would depend on the type.
+If two numbers were passed, the obvious result should be the sum of the two numbers.
+If on the other hand two strings were passed, the function could be implemented in such a way that the result would be the concatenation of the two strings.
+
+The function overloading in HCL is however a bit limited in that functions can not be overloaded with different amounts of parameters, only different parameter types. 
+This is due to the nature of the syntax in HCL where function parameters are not directly coupled to a function. 
+Therefore it would be impossible for the compiler to determine what is parameters and what is functions.
+Consider this example:
+
+\begin{lstlisting}
+1 to 10 sum print
+\end{lstlisting}
+
+Obviously the sum function accepts a list that should be summed. 
+However, if the sum function could potentially also accept a modifier function, that were to alter the result of the sum function, it would be impossible for the compiler to determine whether print was that modifier function, or the sum should be passed as a parameter to the print function.
+
+Function overloading requires nothing special of the user, just declare two functions with the same amount of parameters, but with different types. 
+See listings below for example
+
+\begin{lstlisting}
+var isAwesome = (string name) -> bool {
+	return name equals "HCL" thenElse {true} {false}
+}
+var isAwesome = (num number) -> bool {
+	return number equals 42 thenElse {true} {false}
+}
+
+42 isAwesome // True
+"Java" isAwesome // False
+\end{lstlisting}
+
+The main reasoning for including function overloading is simplicity for the user.
+For instance for printing. 
+The print function should be able to print all basic types, so that the user does not have to worry about the type of what he is about to print.
+
+
+\textbf{Generics}\\
+The language will include simple generics into the language.
```

Comment:
either "include ... in the langauge"
or "introduce ... into the language". currently sounds weird